src/coniql/metrics.py

```diff
@@ -0,0 +1,150 @@
+    "coniql_request_handler_exceptions",
+    "Number of exceptions and GraphQL errors in requests",
+)
+DROPPED_UPDATES = Gauge(
+    "coniql_dropped_updates", "Number of updates dropped in subscriptions"
+)
+ACTIVE_CHANNELS = Gauge("coniql_active_channels", "Number of active channels in aioca")
+
+
+class MetricsExtension(SchemaExtension):
+    def on_operation(self):
+        """Counts the number of GraphQL Queries and Mutations"""
+        REQUESTS.inc({"route": "GraphQL"})
+        yield
+
+
+class SchemaWithMetrics(Schema):
+    """Extended Schema with metrics"""
+
+    def process_errors(
+        self,
+        errors: List[GraphQLError],
+        execution_context: Optional[ExecutionContext] = None,
+    ):
+        """Override to count the number of GraphQL errors"""
+        labels = {"route": "GraphQL"}
+        # Technically there could be multiple GraphQL errors present, but for metric
+        # purposes we really only care that any happened.
+        if len(errors):
+            if (
+                execution_context
+                and execution_context.context
+                and "request" in execution_context.context
+            ):
+                # If we can, add the remote end's address
+                labels.update({"remote": execution_context.context["request"].remote})
+            REQUEST_EXCEPTIONS.inc(labels)
+
+        super().process_errors(errors, execution_context)
+
+
+class MetricsGraphQLTransportWSHandler(GraphQLTransportWSHandler):
+    """Custom override of GraphQLTransportWSHandler to allow adding the @inprogress
+    annotation. Tracks how many subscriptions are currently active."""
+
+    @inprogress(
+        SUBSCRIPTIONS_IN_PROGRESS,
+        labels={"type": f"subscription_{GRAPHQL_TRANSPORT_WS_PROTOCOL}"},
+    )
+    async def handle_request(self):
+        await super().handle_request()
+
+
+class MetricsGraphQLWSHandler(GraphQLWSHandler):
+    """Custom override of GraphQLWSHandler to allow adding the @inprogress
+    annotation. Tracks how many subscriptions are currently active."""
+
+    @inprogress(
+        SUBSCRIPTIONS_IN_PROGRESS,
+        labels={"type": f"subscription_{GRAPHQL_WS_PROTOCOL}"},
+    )
+    async def handle_request(self):
+        await super().handle_request()
+
+
+def update_subscription_metrics(
+    subscription: Subscription, last_dropped_count: int, labels: Dict[str, str]
+) -> int:
+    """Inspect the given Subscription to update the relevant metrics.
+    Must provide the number of dropped updates prevously observed.
+    Returns the current number of dropped updates."""
+    labels.update({"pv": subscription.name})
+    # Ensure only read dropped_callbacks once, in case it is updating in the background
+    current_dropped_count = subscription.dropped_callbacks
+
+    difference = current_dropped_count - last_dropped_count
+
+    # Note that adding 0 will still cause the metric to be published
+    DROPPED_UPDATES.add(labels, difference)
```

Comment:
Why do you have to add the difference, why not set to the current value? Ideally we would avoid passing the value out and having to handle it in the calling code

Reply:
After discussion, we decided to keep using `.add()`. This is because there could be multiple `Subscription` objects connected to one PV, and we don't want to override one `Subscription`'s dropped_updates with the value from another. The current code has the effect of summing all dropped_updates across all `Subscription`s. This may cause double-(or triple-)counting of dropped updates, but this is probably preferable to potentially missing them entirely.

This should be reviewed once this is deployed and we can see some real data.